A69-crl-enhancements.md

```diff
@@ -0,0 +1,222 @@
+
+This approach adds complexity to the user facing API, and a user overriding the provider interface poorly could result in undefined or inefficient behaviors. To counteract this, we will provide generally useful implementations and good documentation on requirements if a user is writing their own implementation.
+
+We explicitly will not be supporting OSCP-style revocation . We want to stay inside the realm of RFC5280 and X509-style CRLs, and OSCP represents a departure from that standard. Notably, the API surface for OSCP is different - for OSCP one makes a call to an external service that returns the revocation status. Rather, we expect to be given CRLs and will check the revocation status in gRPC. In addition, OSCP requires external calls during the handshake, whereas with X509 CRL files everything can done locally. Handshakes are on a path for which performance in critical, so making external calls during them is an approach to avoid.
+
+
+## Implementation 
+### Go
+CRL checking in Go is entirely dependent upon setting the `tls.config.VerifyPeerCertificate` callback function to check CRLs. Currently, the Go implementation overrides this callback and from that calls `CheckChainRevocation(...)` which takes the certificate chain to be checked and the RevocationConfig. It gets the CRL of interest by calling `FetchIssuerCrl()` - we can use our provider in that function. Then it checks the cert chain against the relevant CRLs.
+
+We will add CRL Provider code to the advancedtls module. All CRL related code is currently in the advancedtls module, so this would keep it all in the same place.  Revocation configurations will be added to the existing configuration options in `RevocationConfig`. There will be checks to ensure a user can exclusively use the new provider options or the already-implemented X509 hash directory options. Configuring both is a mistake.
+```go
+// RevocationConfig contains options for CRL lookup.
+type RevocationConfig struct {
+	// RootDir is the directory to search for CRL files.
+	// Directory format must match OpenSSL X509_LOOKUP_hash_dir(3).
+	RootDir string
+	// AllowUndetermined controls if certificate chains with RevocationUndetermined
+	// revocation status are allowed to complete.
+	AllowUndetermined bool
+	// Cache will store CRL files if not nil, otherwise files are reloaded for every lookup.
+	Cache Cache
+
+  // crlProvider and RootDir/Cache cannot both be set
++ CrlProvider crlProvider
+}
+```
+
+The core CrlProvider interface will look as follows. Since x509 is a built in go library, we can safely use it in our API.
+```go
+type CRLProvider interface {
+	// Callers are expected to use the returned value as read-only.
+	CRL(cert x509.Certificate) (*CertificateListExt, error)
+
+	// Close cleans up resources allocated by the Provider.
+	Close()
```

Comment:
@gtcooke94  We don't have `Close` at interface level (I don't think it's relevant for StaticProvider for example) - are you ok to drop it from here?

Reply:
I'm actually curious about this. If we don't have this at the interface level, how does the consumer of the CRLProvider call this func? I don't think casting the interface back to a specific implementation is a good idea. Did I miss anything?

Reply:
@rockspore IMHO this function is relevant only for some of the CRLProviders but not to all of them (for example, StaticProvider, or a provider which reads a directory only once).
However, there is another approach - keep `Close()` on interface level but use a no-op impl if it's not relevant. Such pattern is used in gRPC Go [repo](https://github.com/grpc/grpc-go/blob/70f1a4045da95b93f73b6dbdd7049f3f053c0680/balancer/balancer.go#L388)

Reply:
This pattern is not a nice-to-have but a must. The consumer will store an instance of the interface and has no knowledge about the specific implementation. If `Close` is not defined at the interface level, it won't be able to call it.

Reply:
My point is that not all the needed methods will be present at the interface level. Let's say a consumer develops a `CRLDirectoryFileMaskProvider` with an extra `UpdateFileMaskAndScan` method, which accepts file extensions to be included in the next directory scans. This method won't be part of the interface, so the consumer will need to store the instance (or to work with casting).

Reply:
I think it all depends on who calls the `Close()` and I was thinking this would be handled during the server shutdown. So gRPC needs to be able to stop whatever CRL provider it's given. But if the intent is for gRPC users that created the provider to recycle it, then it probably doesn't matter.

Reply:
We decided to remove this after discussing it in the Golang PR